Add the ability for `Client` to run preload code

bryanwweber opened the pull request:
- [X] Closes #5651 
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

I still need to add more tests here, but posting this for feedback from anyone who wants to chime in! I also need to update the docs!

Also, with respect to the linting failure, should I try to add types for those missing ones that `mypy` identified, or just leave it?

~I suppose this will also need a corresponding PR in `dask/dask` to add the docs, is that correct?~ There is a corresponding docs PR at dask/dask#8679

jrbourbeau commented on distributed/client.py:
```diff
@@ -920,6 +925,18 @@ def __init__(
         for ext in extensions:
             ext(self)
 
+        if not preload:
+            preload = cast("list[str]", dask.config.get("distributed.client.preload"))
+        if not preload_argv:
+            preload_argv = cast(
+                "list[list[str]]", dask.config.get("distributed.client.preload-argv")
+            )
```
I'm curious why these `cast` calls are needed. We have similar logic over in `worker.py` for processing worker preloads

https://github.com/dask/distributed/blob/9fdada1fef36545ef5b560a512d21709d7d46b56/distributed/worker.py#L557

https://github.com/dask/distributed/blob/9fdada1fef36545ef5b560a512d21709d7d46b56/distributed/worker.py#L800-L808

and `mypy` seems to be happy without the `cast` calls over in that module

bryanwweber replied:
Mypy is happy; Pylance complains 😄 I'll remove them.

jrbourbeau commented on distributed/distributed-schema.yaml:
```diff
@@ -525,17 +525,17 @@ properties:
             type: array
             description: |
-              Run custom modules during the lifetime of the scheduler
+              Run custom modules during the lifetime of the nanny
```
Nice catch 👍

jrbourbeau commented on distributed/client.py:
```diff
@@ -1181,6 +1198,9 @@ async def _start(self, timeout=no_default, **kwargs):
         for topic, handler in Client._default_event_handlers.items():
             self.subscribe_topic(topic, handler)
 
+        for preload in self.preloads:
+            await preload.start()
```
In addition, we'll also want to make sure to call the preload teardown methods when the client closes down. Similar to what we do in the worker here

https://github.com/dask/distributed/blob/9fdada1fef36545ef5b560a512d21709d7d46b56/distributed/worker.py#L1615-L1616

bryanwweber replied:
I was wondering why testing for the teardown method didn't work when I added it before 🤦

jrbourbeau commented on distributed/tests/test_preload.py:
```diff
@@ -288,3 +289,70 @@ async def test_web_preload_worker(cleanup, worker_preload):
+
+
+@gen_cluster(nthreads=[])
+async def test_client_preload_text(s: Scheduler):
```
Once preload teardown methods are called when the client closes, we can extend each of these preloads to also have a `dask_teardown(client)` method and make similar `assert`s to what's already here for `dask_setup(client)`

bryanwweber replied:
I only added a teardown to the first two tests because there's no click functionality to be tested for teardowns.

jrbourbeau reviewed (commented):
Thanks @bryanwweber -- overall this looks great. I left a few initial comments, but will give a more thorough look tomorrow

jrbourbeau commented on distributed/client.py:
```diff
@@ -1509,6 +1527,9 @@ async def _close(self, fast=False):
 
             await self.rpc.close()
 
+            for preload in self.preloads:
+                await preload.teardown()
+
```
Let's move this to call the `teardown` methods earlier in the client close process. If I look at when preload `teardown` happens on the scheduler

https://github.com/dask/distributed/blob/d0a93e950d70680c41959c517a68ee6fc0ce2138/distributed/scheduler.py#L4155-L4156

and worker

https://github.com/dask/distributed/blob/d0a93e950d70680c41959c517a68ee6fc0ce2138/distributed/worker.py#L1615-L1616

it happens towards the beginning before we actually start shutting things down. This can useful if, for example, we want the client's teardown method to communicate with the scheduler. It will also make things more consistent with what we already have with scheduler and worker preloads.

jrbourbeau replied:
I'll suggest we move call the `teardown` methods after 

```python
self.status = "closing"
```

above

bryanwweber commented on distributed/tests/test_worker.py:
```diff
@@ -3184,7 +3184,7 @@ async def test_missing_released_zombie_tasks_2(c, s, a, b):
     nthreads=[("", 1)],
     config={"distributed.worker.memory.pause": 0.5},
-    worker_kwargs={"memory_limit": 2 ** 29},  # 500 MiB
+    worker_kwargs={"memory_limit": 2**29},  # 500 MiB
```
It depends on the version of Black. The stable version (22.0 I think?) now puts simple binary operators like this.

jrbourbeau replied:
We use `pre-commit` to manage all code formatting checks (including tool versions). Following https://distributed.dask.org/en/stable/develop.html#code-formatting should make sure what you're using locally matches what's run in CI

jrbourbeau replied:
Ah, I see. See https://github.com/dask/distributed/pull/5782 for an update to `pre-commit` versions

bryanwweber commented:
@jrbourbeau I rebased this onto `main` and got rid of the merge commit when I addressed the `black` comments. I think this addresses all the comments so far now 😄 👍

jrbourbeau reviewed (commented):
Thanks for the updates @bryanwweber. It looks like `mypy` is unhappy with some type annotations

```
distributed/diagnostics/memory_sampler.py:97: error: Item "None" of "Union[rpc, PooledRPCCall, None]" has no attribute "memory_sampler_start"
distributed/diagnostics/memory_sampler.py:105: error: Item "None" of "Union[rpc, PooledRPCCall, None]" has no attribute "memory_sampler_stop"
distributed/diagnostics/memory_sampler.py:112: error: Item "None" of "Union[rpc, PooledRPCCall, None]" has no attribute "memory_sampler_start"
distributed/diagnostics/memory_sampler.py:118: error: Item "None" of "Union[rpc, PooledRPCCall, None]" has no attribute "memory_sampler_stop"
distributed/active_memory_manager.py:446: error: Item "None" of "Union[rpc, PooledRPCCall, None]" has no attribute "amm_handler"
distributed/client.py:786: error: Need type annotation for "futures" (hint: "futures: Dict[<type>, <type>] = ...")
distributed/client.py:787: error: Need type annotation for "refcount"
distributed/client.py:798: error: Need type annotation for "_pending_msg_buffer" (hint: "_pending_msg_buffer: List[<type>] = ...")
distributed/client.py:799: error: Need type annotation for "extensions" (hint: "extensions: Dict[<type>, <type>] = ...")
distributed/client.py:804: error: Need type annotation for "_scheduler_identity" (hint: "_scheduler_identity: Dict[<type>, <type>] = ...")
distributed/client.py:896: error: Need type annotation for "_event_handlers" (hint: "_event_handlers: Dict[<type>, <type>] = ...")
distributed/client.py:2648: error: Item "None" of "Union[rpc, PooledRPCCall, None]" has no attribute "broadcast"
distributed/client.py:3839: error: Item "None" of "Union[rpc, PooledRPCCall, None]" has no attribute "dump_state"
distributed/client.py:3841: error: Item "None" of "Union[rpc, PooledRPCCall, None]" has no attribute "broadcast"
distributed/client.py:4070: error: Item "None" of "Union[rpc, PooledRPCCall, None]" has no attribute "events"
distributed/client.py:4156: error: Item "None" of "Union[rpc, PooledRPCCall, None]" has no attribute "retire_workers"
Found 16 errors in 3 files (checked 254 source files)
```

bryanwweber commented:
@jrbourbeau I reverted the typing in the `__init__` signature for the `preload` and `preload_argv` parameters and the `mypy` issue seems resolved locally. I don't know why adding typing for those two parameters was causing so many problems all over the place, but... it's fixed now 😄

jrbourbeau commented on distributed/client.py:
```diff
@@ -770,6 +773,8 @@ def __init__(
         connection_limit=512,
+        preload=None,
+        preload_argv=None,
         **kwargs,
```
When creating a `LocalCluster` implicitly, these `kwargs` are forwarded to the `LocalCluster` constructor. Can you check is `preload` and `preload_argv` are valid inputs for `LocalCluster`? I'm wondering if there is existing user code that already is passing `preload` / `preload_argv` to `Client`

bryanwweber replied:
Great point. Existing `preload`/`preload_argv` would be passed as `worker_kwargs` into the worker instance, I think. That could potentially be a problem, this change would unexpectedly redirect the `preload` to the `Client` instead. This is really tricky, I'm not sure how to deal with it. I can think of a few options right away:

1. Name these `client_preload`/`client_preload_argv`
   a. Possibly add a deprecation cycle to require `preload`s that are intended for `worker`s or the `scheduler` to be called `worker_preload`/`scheduler_preload`. (There is an existing `scheduler_kwargs` keyword in `LocalCluster`)
2. Hope that no one is passing a `preload` to the worker on a `LocalCluster` (no changes needed)
3. Change `LocalCluster` to not capture all the extra kwargs that it gets, and have `worker_kwargs` be a single argument that can be passed (like `scheduler_kwargs`)

Do you have any suggestions here?

jrbourbeau reviewed (commented):
Thanks @bryanwweber -- one last comment

jrbourbeau commented on distributed/client.py:
```diff
@@ -920,6 +923,12 @@ def __init__(
+        preload = dask.config.get("distributed.client.preload")
+        preload_argv = dask.config.get("distributed.client.preload-argv")
+
+        # These two asserts are for typing
```
```suggestion
```

jrbourbeau reviewed (commented):
Thanks @bryanwweber! One minor comment about an out of date comment, but otherwise this is good to go

jrbourbeau commented:
Also, would you mind merging `main`? There have been several flaky test fixes since this PR was originally opened

bryanwweber commented:
@jrbourbeau 🤦 Dunno how I missed that comment, thanks for the spot. Should be all set now!

jrbourbeau reviewed (approved):
Thanks @bryanwweber -- this is in
